README.md
# webdev

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2020 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-11-24
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-11-24
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/webdev/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/webdev/compare/0.1.0...0.1.1


src/webdev/item/view.clj
(ns webdev.item.view
  (:require [hiccup.page :refer [html5]]
            [hiccup.core :refer [html h]]))

(defn new-item []
  (html
    [:form.form-horizontal
     {:method "POST" :action "/items"}
     [:div.form-group
      [:label.control-label.col-sm-2 {:for :name-input}
       "Name"]
      [:div.col-sm-10
       [:input#name-input.form-control
        {:name        :name
         :placeholder "Name"}]]]
     [:div.form-group
      [:label.control-label.col-sm-2 {:for :desc-input}
       "Description"]
      [:div.col-sm-10
       [:input#desc-input.form-control
        {:name        :description
         :placeholder "Description"}]]]
     [:div.form-group
      [:div.col-sm-offset-2.col-sm-10
       [:input.btn.btn-primary
        {:type  :submit
         :value "New item"}]]]]))

(defn delete-item-form [id]
  (html
    [:form
     {:method "POST" :action (str "/items/" id)}
     [:input {:type  :hidden
              :name  "_method"
              :value "DELETE"}]
     [:div.btn-group
      [:input.btn.btn-danger.btn-xs
       {:type  :submit
        :value "Delete"}]]]))

(defn update-item-form [id checked]
  (html
    [:form
     {:method "POST" :action (str "/items/" id)}
     [:input {:type  :hidden
              :name  "_method"
              :value "PUT"}]
     [:input {:type  :hidden
              :name  "checked"
              :value (if checked "false" "true")}]
     [:div.btn-group
      [:button.btn.btn-primary.btn-xs
       {:type :submit}
       (if checked "DONE" "TODO")]]]))

(defn items-page [items]
  (html5 {:lang :en}
         [:head
          [:title "BA List"]
          [:meta {:name    :viewport
                  :content "width=device-width, initial-scale=1.0"}]
          [:link {:href "/bootstrap/css/bootstrap.min.css"
                  :rel  :stylesheet}]]
         [:body
          [:div.container
           [:div.row
            [:h1 "My items"]
            (if (seq items)
              [:table.table.table-striped
               [:thead
                [:tr
                 [:th.col-sm-2]
                 [:th.col-sm-2]
                 [:th "Name"]
                 [:th "Description"]]]
               [:tbody
                (for [i items]
                  [:tr
                   [:td (delete-item-form (:id i))]
                   [:td (update-item-form (:id i) (:checked i))]
                   [:td (h (:name i))]
                   [:td (h (:description i))]])]]
              [:div.col-sm-offset-1 "There are no items."])
            [:div.col-sm-6
             [:h2 "Create a new item"]
             (new-item)]]]
          [:script {:src "https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"}]
          [:script {:src "/bootstrap/js/bootstrap.min.js"}]]))



src/webdev/item/model.clj
(ns webdev.item.model
  (:require [clojure.java.jdbc :as db]))

(defn create-table! [db]
  (db/execute! db
               ["CREATE EXTENSION IF NOT EXISTS \"uuid-ossp\""])
  (db/execute! db
               ["CREATE TABLE IF NOT EXISTS items
                (id UUID PRIMARY KEY DEFAULT uuid_generate_v4(),
                 name TEXT NOT NULL,
                 description TEXT NOT NULL,
                 checked BOOLEAN NOT NULL DEFAULT FALSE,
                 date_created TIMESTAMPTZ NOT NULL DEFAULT now())"]))

(defn create-item! [db name description]
  (:id (first (db/query db
                        ["INSERT INTO items (name, description)
                          VALUES (?, ?)
                          RETURNING id"
                         name
                         description]))))

(defn update-item! [db id checked]
  (= [1] (db/execute! db
                      ["UPDATE items
                             SET checked = ?
                             WHERE id = ?"
                       checked
                       id])))

(defn delete-item! [db id]
  (= [1] (db/execute! db
                      ["DELETE FROM items
                             WHERE id = ?"
                       id])))

(defn read-items! [db]
  (db/query db
            ["SELECT id, name, description, checked, date_created
             FROM items
             ORDER BY date_created"]))


src/webdev/item/handler.clj
(ns webdev.item.handler
  (:require [webdev.item.model :as item]
            [webdev.item.view :as view])
  (:import (java.util UUID)))

(defn handle-index-items [req]
  (let [db (:webdev/db req)
        items (item/read-items! db)]
    {:status  200
     :headers {}
     :body    (view/items-page items)}))

(defn handle-create-item [req]
  (let [db (:webdev/db req)
        name (get-in req [:params "name"])
        description (get-in req [:params "description"])
        item-id (item/create-item! db name description)]
    {:status  302
     :headers {"Location" "/items"}
     :body    ""}))

(defn handle-delete-item [req]
  (let [db (:webdev/db req)
        item-id (UUID/fromString (:item-id (:route-params req)))
        exists? (item/delete-item! db item-id)]
    (if exists?
      {:status  302
       :headers {"Location" "/items"}
       :body    ""}
      {:status 404
       :header {}
       :body   "List not found"})))

(defn handle-update-item [req]
  (let [db (:webdev/db req)
        item-id (UUID/fromString (:item-id (:route-params req)))
        checked (get-in req [:params "checked"])
        exists? (item/update-item! db item-id (= "true" checked))]
    (if exists?
      {:status  302
       :headers {"Location" "/items"}
       :body    ""}
      {:status 404
       :header {}
       :body   "Item not found"})))

src/webdev/core.clj
(ns webdev.core
  (:require [ring.adapter.jetty :as jetty]
            [ring.middleware.reload :refer [wrap-reload]]
            [ring.middleware.params :refer [wrap-params]]
            [ring.middleware.resource :refer [wrap-resource]]
            [ring.middleware.file-info :refer [wrap-file-info]]
            [compojure.core :refer [defroutes ANY GET PUT POST DELETE]]
            [compojure.route :refer [not-found]]
            [ring.handler.dump :refer [handle-dump]]
            [webdev.item.model :as items]
            [webdev.item.handler :as handle-index-items]))

(def db (or
          (System/getenv "DATABASE_URL")
          "jdbc:postgresql://postgres:1234@postgres.local:5432/webdev"))

(defn greet [req]
  {:status  200
   :body    "Hello World!"
   :headers {}})

(defn goodbye [req]
  {:status  200
   :body    "Goodbye cruel world!"
   :headers {}})

(defn about [req]
  {:status  200
   :body    "I'm a Clojure developer :)"
   :headers {}})

(defn yo [req]
  (let [name (get-in req [:route-params :name])]
    {:status  200
     :body    (str "Yo " name "!")
     :headers {}}))

(def ops
  {"+" +
   "-" -
   "*" *
   ":" /})

(defn calc [req]
  (let [a (Integer. (get-in req [:route-params :a]))
        b (Integer. (get-in req [:route-params :b]))
        op (get-in req [:route-params :op])
        f (get ops op)]
    (if f
      {:status  200
       :body    (str (f a b))
       :headers {}}
      {:status  404
       :body    "Unknown operation"
       :headers {}})))

(defroutes routes
           (GET "/" [] greet)
           (GET "/goodbye" [] goodbye)
           (GET "/about" [] about)
           (ANY "/request" [] handle-dump)
           (GET "/yo/:name" [] yo)
           (GET "/calc/:a/:op/:b" [] calc)

           (GET "/items" [] handle-index-items/handle-index-items)
           (POST "/items" [] handle-index-items/handle-create-item)
           (DELETE "/items/:item-id" [] handle-index-items/handle-delete-item)
           (PUT "/items/:item-id" [] handle-index-items/handle-update-item)

           (not-found "Page not found"))

(defn wrap-db [handler]
  (fn [req]
    (handler (assoc req :webdev/db db))))

(defn wrap-server [handler]
  (fn [req]
    (assoc-in (handler req) [:headers "Server"] "BA List")))

(def sim-methods
  {"PUT"    :put
   "DELETE" :delete})

(defn wrap-simulated-methods [handler]
  (fn [req]
    (if-let [method (and (= :post (:request-method req))
                         (sim-methods (get-in req [:params "_method"])))]
      (handler (assoc req :request-method method))
      (handler req))))

(def app
  (wrap-server
    (wrap-file-info
      (wrap-resource
        (wrap-db
          (wrap-params
            (wrap-simulated-methods
              routes)))
        "static"))))

(defn -main [port]
  (items/create-table! db)
  (jetty/run-jetty app
                   {:port (Integer. port)}))

(defn -dev-main [port]
  (items/create-table! db)
  (jetty/run-jetty (wrap-reload #'app)
                   {:port (Integer. port)}))

;; TODO
;; Make homepage HTML, add a picture and show links to all GET routes
;; Rework the system to support multiple lists and multiple items each